pkg/schemacompat/schemacompat.go

```diff
@@ -0,0 +1,374 @@
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// EnsureStructuralSchemaCompatibility compares a new structural schema to an existing one, to ensure that the existing
+// schema is a sub-schema of the new schema. In other words that means that all the documents validated by the existing schema
+// will also be validated by the new schema, so that the new schema can be considered backward-compatible with the existing schema.
+// If it's not the case, errors are reported for each incompatible schema change.
+//
+// PLEASE NOTE that the implementation is incomplete (it's ongoing work), but still consistent:
+// if some Json Schema elements are changed and the comparison of this type of element is not yet implemented,
+// then an incompatible change error is triggered explaining that the compatison on this element type is not supported.
+// So there should never be any case when a schema is considered backward-compatible while in fact it is not.
+//
+// If the narrowExisting argument is true, then the LCD (Lowest-Common-Denominator) between existing schema and the new schema
+// is built (when possible), and returned if no incompatible change was detected (like type change, etc...).
+// If the narrowExisting argument is false, the existing schema is untouched and no LCD schema is calculated.
+//
+// In either case, when no errors are reported, it is ensured that either the existing schema or the calculated LCD
+// is a sub-schema of the new schema.
+func EnsureStructuralSchemaCompatibility(fldPath *field.Path, existing, new *apiextensionsv1.JSONSchemaProps, narrowExisting bool) (lcd *apiextensionsv1.JSONSchemaProps, errors utilerrors.Aggregate) {
+	var newInternal, existingInternal apiextensions.JSONSchemaProps
+	apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(existing, &existingInternal, nil)
+	apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(new, &newInternal, nil)
+	newStrucural, err := schema.NewStructural(&newInternal)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+
+	existingStructural, err := schema.NewStructural(&existingInternal)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+
+	lcdStructural := existingStructural.DeepCopy()
+	errs := lcdForStructural(fldPath, existingStructural, newStrucural, lcdStructural, narrowExisting)
+	if len(errs) > 0 {
+		return nil, errs.ToAggregate()
+	}
+	serialized, err := json.Marshal(lcdStructural.ToGoOpenAPI())
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+	var jsonSchemaProps apiextensionsv1.JSONSchemaProps
+	json.Unmarshal(serialized, &jsonSchemaProps)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+	return &jsonSchemaProps, nil
+}
+
+func checkTypesAreTheSame(fldPath *field.Path, existing, new *schema.Structural) (errorList field.ErrorList) {
```

Comment:
Can this just `return err` too? It looks like you could just `return field.Invalid(...)`, which satisfies the `error` interface.

Reply:
Well, in fact I'm not really interested at returning the `error` interface, but more a `*field.Error`.

The logic here is that top-level function (`EnsureStructuralSchemaCompatibility`) returns an Aggregate of errors which can be either `*field.Error` for every error related to the content of the schema, or a generic `error` for errors that occured outside of proper schema comparison (structural schema parse, etc ...).
That's why all schema comparison function share the same return value, which is a list of `field.*Error`. I agree that in some cases we know we only return one error (especially in the case that are not yet implemented), but in full generality comparison at a given level might return several errors. Let's even take the case of integers: as soon as they are implemented, the comparison function might return 2 incompatibility errors: on related to intervals, the other related to `multiple-of`.

So here I can remove the `ErrorList` and return only 1 `*field.Error`, but it's really one of the few case where we're sure we'll never return several errors.

Reply:
Quick follow up: when switching the return value here to a `*field.Error`, I'll have to test, at all call sites, whether the returned value is nil, to avoid adding a nil error in the error list.
That doesn't see to make the code cleaner to me.